File: practice/mathematics/fundamentals/even_odd_query/even_odd_query.hs
module Main where

import qualified Data.Vector as V

solve :: V.Vector Int -> [Int] -> String
solve ar [x, y]
    | x == y = solve'  ar x
    | x /= y = solve'' ar x
    where
        solve'  ar x
            | ar V.! (x - 1) `mod` 2 == 1 = "Odd"
            | otherwise                   = "Even"
        solve'' ar x
            | ar V.! (x - 1) `mod` 2 == 0 && ar V.! x /= 0 = "Even"
            | otherwise                                    = "Odd"

main :: IO ()
main = do
    n <- readLn :: IO Int
    ta <- getLine
    q <- readLn :: IO Int
    
    let a = map (read :: String -> Int) $ words ta
        ar = V.fromList a
    
    t <- getContents
    putStr $ unlines $ map (solve ar) $ map (map (read :: String -> Int) . words) $ lines t

File: practice/mathematics/fundamentals/even_odd_query/even_odd_query.cpp
#include <vector>
#include <iostream>

using namespace std;

int main()
{
    int n, x, y, q;
    
    cin >> n;
    
    vector<int> a(n);
    for (int pos = 0; pos < n; ++pos)
    {
        cin >> a[pos];
    }
    
    cin >> q;
    while (q--)
    {
        cin >> x >> y;
        if (x == y)
        {
            if (a[x - 1] % 2 == 1)
            {
                cout << "Odd" << endl;
            }
            else
            {
                cout << "Even" << endl;
            }
        }
        else
        {
            if (a[x - 1] % 2 == 0 && a[x] != 0)
            {
                cout << "Even" << endl;
            }
            else
            {
                cout << "Odd" << endl;
            }
        }
    }
    
    return 0;
}

